cts/cts-cli.in

```diff
@@ -152,15 +456,133 @@ def setup_path(opts_path):
         os.environ["PCMK_schema_directory"] = BuildOptions.SCHEMA_DIR
 
 
+def _run_one(valgrind, r):
+    """Run and return a TestGroup object."""
+    # See comments in run_regression_tests.
+    r.run(valgrind=valgrind)
+    return r
+
+
+def run_regression_tests(regs, jobs, valgrind=False):
+    """Run the given tests and return the modified objects."""
+    executed = []
+
+    with Pool(processes=jobs) as pool:
+        # What we really want to do here is:
+        #     pool.map(lambda r: r.run(),regs)
+        #
+        # However, multiprocessing uses pickle somehow in its operation, and python
+        # doesn't want to pickle a lambda (nor a nested function within this one).
+        # Thus, we need to use the _run_one wrapper at the file level just to call
+        # run().  Further, if we don't return the modified object from that and then
+        # return the list of modified objects here, it looks like the rest of the
+        # program will use the originals, before this was ever run.
+        executed = pool.map(partial(_run_one, valgrind), regs)
+
+    return executed
+
+
+def results(regs, save, verbose):
+    """Print the output from each regression test, returning the number whose output differs."""
+    output_differs = 0
+
+    if verbose:
+        print("\n\nResults")
+
+    for r in regs:
+        r.write()
+
+        r.diff()
+        if not r.identical:
+            output_differs += 1
+
+        if save:
+            dest = "%s/cli/regression.%s.exp" % (test_home, r.name)
+            copyfile(r.results_file, dest)
+
+    return output_differs
+
+
+def summary(regs, output_differs, verbose):
+    """Print the summary output for the entire test run."""
+    test_failures = 0
+    test_successes = 0
+
+    for r in regs:
+        test_failures += r.failures
+        test_successes += r.successes
+
+    print("\n\nSummary")
+
+    # First, print all the Passed/Failed lines from each Test run.
+    for r in regs:
+        print("\n".join(r.summary))
+
+    # Then, print information specific to each result possibility.  Basically,
+    # if there were failures then we print the output differences, leave the
+    # failed output files in place, and exit with an error.  Otherwise, clean up
+    # anything that passed.
+    if test_failures > 0 and output_differs > 0:
+        print("%d test failed; see output in:" % test_failures)
```

Comment:
tests (or even better check singular vs plural for here and the rest of the function)

Reply:
Added a solution that's potentially overkill, but potentially more generically useful.